hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactingMemStore.java

```diff
@@ -0,0 +1,565 @@
+        // compaction is in progress
+        compactor.startCompact();
+      }
+    } catch (IOException e) {
+      LOG.warn("Unable to run memstore compaction. region "
+          + getRegionServices().getRegionInfo().getRegionNameAsString()
+          + "store: "+ Bytes.toString(getFamilyName()), e);
+    }
+  }
+
+  private byte[] getFamilyName() {
+    return store.getFamily().getName();
+  }
+
+  private ExecutorService getPool() {
+    RegionServerServices rs = getRegionServices().getRegionServerServices();
+    if(rs==null) return null;
+    return rs.getExecutorService();
+  }
+
+  private boolean shouldFlushInMemory() {
+    if(getActive().getSize() > 0.9*flushSizeLowerBound) { // size above flush threshold
+      return (allowCompaction.get() && !inMemoryFlushInProgress.get());
+    }
+    return false;
+  }
+
+  /**
+   * The request to cancel the compaction asynchronous task (caused by in-memory flush)
+   * The compaction may still happen if the request was sent too late
+   * Non-blocking request
+   */
+  private void stopCompact() {
+    if (inMemoryFlushInProgress.get()) {
+      compactor.stopCompact();
+      inMemoryFlushInProgress.set(false);
+    }
+  }
+
+  private void pushActiveToPipeline(MutableSegment active) {
+    if (!active.isEmpty()) {
+      long delta = DEEP_OVERHEAD_PER_PIPELINE_ITEM - DEEP_OVERHEAD;
+      active.setSize(active.getSize() + delta);
+      pipeline.pushHead(active);
+      resetCellSet();
+    }
+  }
+
+  private void pushTailToSnapshot() {
+    ImmutableSegment tail = pipeline.pullTail();
+    if (!tail.isEmpty()) {
+      setSnapshot(tail);
+      long size = getSegmentSize(tail);
+      setSnapshotSize(size);
+    }
+  }
+
+  private RegionServicesForStores getRegionServices() {
+    return regionServices;
+  }
+
+
+
+  /*----------------------------------------------------------------------
+  * The in-memory-flusher thread performs the flush asynchronously.
+  * There is at most one thread per memstore instance.
+  * It takes the updatesLock exclusively, pushes active into the pipeline,
+  * and compacts the pipeline.
+  */
+  private class InMemoryFlushWorker extends EventHandler {
+
+    public InMemoryFlushWorker () {
+      super(getRegionServices().getRegionServerServices(), EventType.RS_IN_MEMORY_FLUSH_AND_COMPACTION);
+    }
+
+    @Override
+    public void process() throws IOException {
+      flushInMemory();
+    }
+  }
+
+  /**
+   * The ongoing MemStore Compaction manager, dispatches a solo running compaction
+   * and interrupts the compaction if requested.
+   * The MemStoreScanner is used to traverse the compaction pipeline. The MemStoreScanner
+   * is included in internal store scanner, where all compaction logic is implemented.
+   * Threads safety: It is assumed that the compaction pipeline is immutable,
+   * therefore no special synchronization is required.
+   */
+  private class MemStoreCompactor {
```

Comment:
Totally agree with Eddie, I suggest to move it to separate file as CompactionPipieline is. The MemStoreCompactor class is also going to be significantly extended with CellBlocks addition.